### PR TITLE
FIX: FulltextFilter did not work and was not usable

### PR DESCRIPTION
**Pull request by tractorcow**
This is a rebase of https://github.com/silverstripe/silverstripe-framework/pull/3179, including a suggested fallback to enable single-column fulltext specification.

Any feedback @micmania1?

### CODE REVIEW

**Comment by micmania1** [comment]
Any more on this?

**Comment by tractorcow** [comment]
Sorry, I have been neglecting my old PR. I'll make an update soon(ish).

File: search/filters/SearchFilter.php [hunk 1]
```diff
@@ -178,7 +183,7 @@ public function getDbName() {
 			return '"' . implode('"."', $parts) . '"';
 		}
 		
-		return "\"$candidateClass\".\"$this->name\"";
+		return "\"{$candidateClass}\".\"{$this->name}\"";
```
Comment (tractorcow) [inline search/filters/SearchFilter.php]: Well, $var replacement doesn't work with single quotes, so I've replaced this line and used a sprintf instead. Sound better?

**Comment by tractorcow** [comment]
Rebased. :D

**Comment by micmania1** [comment]
Yawn

**Comment by tractorcow** [comment]
ahaha, we need more good anigifs on here ;P

File: tests/search/FulltextFilterTest.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		$this->assertEquals(
+			array("MATCH (\"FulltextDataObject\".\"ColumnA\") AGAINST ('SilverStripe')"),
+			$query3->query()->getWhere()
+		);
+	}
+	
+}
+
+
+class FulltextDataObject extends DataObject {
```
Comment (dhensby) [inline tests/search/FulltextFilterTest.php]: `implements TestOnly` please

File: search/filters/SearchFilter.php [hunk 3]
```diff
@@ -166,6 +166,11 @@ public function getDbName() {
 			return $this->name;
 		}
 		
+		// Ensure that we're dealing with a DataObject.
+		if (!is_subclass_of($this->model, 'DataObject')) {
+			throw new Exception("Model supplied to " . get_class($this) . " should be an instance of DataObject.");
```
Comment (dhensby) [inline search/filters/SearchFilter.php]: Is there not more descriptive `Exception` class to use?

Comment (micmania1) [inline search/filters/SearchFilter.php]: This PR is from before the Internet existed so probably not.

**Comment by tractorcow** [comment]
Rebased and updated based on feedback. It only took all of the first half of this year to do.

**Comment by tractorcow** [comment]
@dhensby :D

**Comment by tractorcow** [comment]
Ta!

**Comment by uniun** [comment]
And it still does not work! :D

Look at lines [FulltextFilter.php#L61](https://github.com/silverstripe/silverstripe-framework/blob/3.2/search/filters/FulltextFilter.php#L61) and [FulltextFilter.php#L66](https://github.com/silverstripe/silverstripe-framework/blob/3.2/search/filters/FulltextFilter.php#L66). Columns need to be with their table name or it may produce duplicate column name error.

I wrote a function to prepare columns for the query:

	private function prepareColumns($columns) {
		$parts = explode(',', $columns);
		array_walk($parts, function(&$col, $key, $model) {
			$name = trim($col, " \t\n\r\0\x0B\"");
			$class = ClassInfo::table_for_object_field($model, $name);
			$col = sprintf('"%s"."%s"', $class, $name);
		}, $this->model);
		return implode(',', $parts);
	}

**Comment by tractorcow** [comment]
Good spotting @uniun ; Let's make the necessary fix. :D

In this case all columns are from the same table, anyway (since indexes can't span multiple tables) so it should be pretty simple.

**Comment by tractorcow** [comment]
Logged at https://github.com/silverstripe/silverstripe-framework/issues/4483

**Pull request merged**